app/js/mongodb.js

```diff
@@ -0,0 +1,34 @@
+const Settings = require('settings-sharelatex')
+const { MongoClient, ObjectId } = require('mongodb')
+
+const clientPromise = MongoClient.connect(Settings.mongo.url)
+const dbPromise = clientPromise.then((client) => client.db())
+
+async function getCollection(name) {
+  return (await dbPromise).collection(name)
+}
```

Comment:
Since we're already waiting for the db before startup, can we assume that the connection is established when someone interacts with the db instead of checking it every time? It would look something like this:
```js
let db = null
async function connect() {
  const client = await MongoClient.connect(Settings.mongo.url)
  db = client.db()
}
const dbPromise = connect()

async function waitForDb() {
  await dbPromise
}

...

module.exports = { db, ... }
```
Then, calling code could simply write:
```js
const { db } = require('./mongodb')
await db.collection('projects').findOne(...)
```
For this to work with tests, tests also need to wait for the db, but this is easy:
```js
before(waitForDb)
```

Reply:
This won't work as proposed. The `db` will get exported by value -- as `null` --, not by reference [1]. Other modules will import `db=null` then. We could work around that: 1) assigning db on the exports object after connecting 2) importing all of our mongodb module in every calling module 3) accessing the db as mongodb.db.collection(). Given that `db` was the final object previously, this may confuse other developers who want to keep using the old, partial import. I think the current implementation is cleaner/simpler to work with. With the refactor to async/await the call-sites can use `const entry = (await myCollectionPromise).findOne({})`.

---
[1]
```js
> let db
undefined
> const exports = {db}
undefined
> db = 'client.db()'
'client.db()'
> exports
{ db: undefined }
```

Reply:
ah. right. good point. It could also be hidden in a synchronous `getDb()` function, but at this point, I'm splitting hairs :smile:

Reply:
We could also adapt https://github.com/overleaf/docstore/blob/dcd5bf1d6bfaff7f8f47bcfe339c7ade9601d858/test/acceptance/js/ArchiveDocsTests.js#L34-L41 in the app code -- replacing `before` with `waitForDb.then`. This may give a similar feeling as with mongojs. Shall we open an RFC ticket?

Reply:
Good idea. It looks like it would avoid the import problems you described. It also has the benefit of having a more sensible behaviour if you make a typo in the collection name: crash instead of allowing Mongo to create a new collection. I don't see any downside. I think you could go ahead and implement this if you'd like.

Reply:
https://github.com/overleaf/chat/pull/57 I'll port it over to the other services once it landed.